### ISSUE TITLE
OrientDB. Connection leaks or just not closing connections

### ISSUE THREAD

**Issue by mikayeln** [orientechnologies/orientdb#6488]
## OrientDB Version, operating system, or hardware.
- v2.0 SNAPSHOT[ ] - .18[ ] .17[ ] .16[ ] .15[ ] .14[ ] .13[ ] .12[ ] .11[ ] .10[ ] .9[ ] .8[ ] .7[ ] .6[ ] .5[ ] .4[ ] .3[ ] .2[ ] .1[ ] .0[ ]
- v2.1 SNAPSHOT[x] - .19[x],.16[ ] .15[ ] .14[ ] .13[ ] .12[ ] .11[ ] .10[ ] .9[ ] .8[ ] .7[ ] .6[ ] .5[ ] .4[ ] .3[ ] .2[ ] .1[ ] .0[ ]
- v2.2 SNAPSHOT[ ] - .rc1[ ] .beta2[ ] .beta1[ ]

## Operating System
- [x] Linux
- [ ] MacOSX
- [ ] Windows
- [ ] Other Unix
- [ ] Other, name? 

We have a large multithreaded Java EE application running on Wildfly 8.
We are using OrientDB 2.1.19.

And we have some problems with the connection leaks. At some point orient server stops responding and all threads working with db stuck on retrieving new connection.

Configuration is following:

    OGlobalConfiguration.CLIENT_CONNECT_POOL_WAIT_TIMEOUT.setValue(5000);
    OGlobalConfiguration.CLIENT_DB_RELEASE_WAIT_TIMEOUT.setValue(5000);
	OGlobalConfiguration.CLIENT_CHANNEL_MAX_POOL.setValue(1000);
	OGlobalConfiguration.DB_POOL_MIN.setValue(100);
	OGlobalConfiguration.DB_POOL_MAX.setValue(5000);
	OGlobalConfiguration.STORAGE_LOCK_TIMEOUT.setValue(5000);
	OGlobalConfiguration.DB_POOL_IDLE_TIMEOUT.setValue(5000);
	OGlobalConfiguration.DB_POOL_IDLE_CHECK_DELAY.setValue(1000);

OPartitionedDatabasePool we're getting thru OPartitionedDatabasePoolFactory

    poolFactory = new OPartitionedDatabasePoolFactory();
    documentPool = poolFactory.get(orientDBPath, username, password);

Then depending on our needs we are using ODatabaseDocumentTx or OObjectDatabaseTx using 

    documentPool.acquire();

Acquired ODatabaseDocumentTx or OObjectDatabaseTx we are passing to so called executor, which executes Runnable or Callable.

    public void run(ORunnable<DB> oRunnable) {
        // db is the private member of executor containing acquired db.
        try {
            //...
			oRunnable.run(db);			
            //...
		} catch (Exception e) {
			
		} finally {
			if(!db.isClosed()) {
				db.close();
			}
		}
    }

As you can see we are closing db in finally section, also we are fully detaching all documents recieved from db. But in some cases number of connections to DB is not dropping and after that we're getting following exception

        2016-07-28 23:10:45,957 FINE  [com.orientechnologies.orient.client.remote.ORemoteConnectionManager] (default task-46) Network connection pool is receiving a closed connection to reuse: discard it
    2016-07-28 23:10:45,957 FINE  [com.orientechnologies.orient.client.remote.ORemoteConnectionManager] (default task-46) Cannot unlock connection lock: java.lang.IllegalMonitorStateException
     at java.util.concurrent.locks.ReentrantLock$Sync.tryRelease(ReentrantLock.java:155) [rt.jar:1.7.0_13]
     at java.util.concurrent.locks.AbstractQueuedSynchronizer.release(AbstractQueuedSynchronizer.java:1260) [rt.jar:1.7.0_13]
      at java.util.concurrent.locks.ReentrantLock.unlock(ReentrantLock.java:460) [rt.jar:1.7.0_13]
      at com.orientechnologies.common.concur.lock.OAdaptiveLock.unlock(OAdaptiveLock.java:123) [orientdb-core-2.1.19.jar:2.1.19]
      at com.orientechnologies.orient.enterprise.channel.binary.OChannelBinaryAsynchClient.unlock(OChannelBinaryAsynchClient.java:371) [orientdb-enterprise-2.1.19.jar:2.1.19]
      at com.orientechnologies.orient.client.remote.ORemoteConnectionManager.remove(ORemoteConnectionManager.java:128) [orientdb-client-2.1.19.jar:2.1.19]
      at com.orientechnologies.orient.client.remote.ORemoteConnectionManager.release(ORemoteConnectionManager.java:119) [orientdb-client-2.1.19.jar:2.1.19]
      at com.orientechnologies.orient.client.remote.OStorageRemote.endResponse(OStorageRemote.java:1643) [orientdb-client-2.1.19.jar:2.1.19]
      at com.orientechnologies.orient.client.remote.OStorageRemote.command(OStorageRemote.java:1240) [orientdb-client-2.1.19.jar:2.1.19]
      at com.orientechnologies.orient.client.remote.OStorageRemoteThread.command(OStorageRemoteThread.java:453) [orientdb-client-2.1.19.jar:2.1.19]
      at com.orientechnologies.orient.core.sql.query.OSQLQuery.run(OSQLQuery.java:72) [orientdb-core-2.1.19.jar:2.1.19]
      at com.orientechnologies.orient.core.sql.query.OSQLSynchQuery.run(OSQLSynchQuery.java:85) [orientdb-core-2.1.19.jar:2.1.19]
      at com.orientechnologies.orient.core.query.OQueryAbstract.execute(OQueryAbstract.java:33) [orientdb-core-2.1.19.jar:2.1.19]
      at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.query(ODatabaseDocumentTx.java:717) [orientdb-core-2.1.19.jar:2.1.19]

Here is the piece of code for the exception above

    ODocument storedSession = orient.onDocuments().call(new ODocCallable<ODocument>() {
						@Override
						public ODocument call(ODatabaseDocumentTx db) {
							try {
								List<ODocument> list = db.query(new OSQLSynchQuery<ODocument>("select from " + CLAZZ + " where storedSession_sessionID = ?"), sessionId);

								if (list.isEmpty()){
									return null;
								}
								return documentUnpin(list.get(0));
							} catch (Exception e) {
								// PersistentSession may not be in DB on request (for example in workspaces)
							}
							return null;
						}
					});

Any suggestions how to improve the situation with connections?
Thanks.

**Comment by tglman** [comment]
hi @mikayeln,

checking it

**Comment by tglman** [comment]
hi @mikayeln,

Is the database acquire from the pool by the same thread that use it and release it ?

Regards

**Comment by mikayeln** [comment]
Hi @tglman,

Yes. The database acquired from the pool will be released by the same thread, which has received the database.

**Comment by mikayeln** [comment]
Under what conditions acquired connection will not be released?

**Comment by tglman** [comment]
hi @mikayeln,

actually the release should be done by the same thread that did the actquire:
ex:
```
public void run(ORunnable<DB> oRunnable) {
    // db is the private member of executor containing acquired db.
    DB db= pool.acquire();
    try {
        //...
        oRunnable.run(db);            
        //...
    } catch (Exception e) {

    } finally {
        if(!db.isClosed()) {
            db.close();
        }
    }
}
```

otherwhise could be there some issues on releasing.

**Comment by mikayeln** [comment]
Hi @tglman ,

Thanks for the suggestion. We foolwed it, also we swithed orient version to 2.2.8.
And we faced another interesting situation.

At some point OClientConnectionManager kept to itself ~80000 (sometimes more) active sockets. And as a result all requests're rejected. So the question is why sockets are still open? SO timeout is set to 5000ms. storageKeepOpen = false. cleanDelay = 2000ms.

**Comment by tglman** [comment]
hi @mikayeln,

We did a few fix on this topic in the last few days, some are already included in 2.2.9 and other will be released in 2.2.10, i'm quite confident that with 2.2.10 this is going to be sorted out.

the 2.2.10 is going to be at most in ~2 weeks.

Regards

**Comment by tglman** [comment]
hi, 

This was fixed and released, closing.

bye
